Add written documentation for the ctc error types with example graphs (#178)
Janelia-Trackathon-2023/traccuracy: docs/source/track_errors/ctc.rst

            axis.set_xlabel("Time")

        handles = [
            Patch(color=gt_color),
            Patch(color=pred_color),
            Patch(color=mapping_color),
            Patch(color='orange'),
            Patch(color='purple')
        ]
        labels = ["Ground Truth", "Prediction", "Mapping", "Edge Annotations", "Node Annotations"]
        ax[-1].legend(handles=handles, labels=labels, loc="center")
        ax[-1].set_axis_off()
        plt.tight_layout()
        fig.suptitle(title, y=1.05)

Nodes
-----

True Positives
^^^^^^^^^^^^^^

A true positive node is one that is matched to only one node in the predicted graph. Additionally, the predicted node is not matched to any other node in the ground truth. True positive nodes are annotated on both the ground truth and the predicted graph.

False Positives
^^^^^^^^^^^^^^^

False positive nodes are annotated on the predicted graph and correspond to a predicted node without a match in the ground truth graph.

False Negatives
^^^^^^^^^^^^^^^

False negative nodes are annotated on the ground truth graph and correspond to a ground truth node without a match in the predicted graph.

Non-Split
^^^^^^^^^

> Non-split nodes are annotated on the predicted graph and correspond to a node in the prediction that has been matched to two nodes in the ground truth graph.

Reviewer: Technically speaking a GT node could consist of more than two predicted nodes if the undersegmentation is severe, so tweaking this to include that possibility.

```suggestion
Non-split nodes are annotated on the predicted graph and correspond to a node in the prediction that has been matched to more than one node in the ground truth graph.
```

.. jupyter-execute::
    :hide-code:

    plot_matched([ex_graphs.two_to_one(t) for t in [0, 1]], [{4: "NS"}, {5: "NS"}], "Non-Split Nodes")

Edges
-----

False Positives
^^^^^^^^^^^^^^^

False positive edges are annotated on the predicted graph. An edge is considered a false positive if both nodes are true positive nodes, but the edge does not match to any edge in the ground truth graph. In the example below, edge (4, 8) is a false positive.

.. jupyter-execute::
    :hide-code:

>     # plot_matched([ex_graphs.crossover_edge()], [{}], "")

Reviewer: Looking at the readthedocs preview, this example doesn't show up, only the ones below it do. Possibly a syntax error here?

Author: I added this function on another branch (I think the track-errors branch). When I merge that set of PRs in I'll uncomment out this line so that the plot shows up on the docs.

False Negatives
^^^^^^^^^^^^^^^

False negative edges are annotated on the ground truth graph. An edge is considered a false negative if:

> 1. Either node is annotated as false negative nodes

Reviewer: ```suggestion
1. Either node is annotated as a false negative node
```

.. jupyter-execute::
    :hide-code:

    plot_matched([ex_graphs.fn_node_matched(0)], [{1: "FN", (1, 2): "FN"}], "")

2. The corresponding edge in the predicted graph does not exist between two true positive nodes

.. jupyter-execute::
    :hide-code:

    plot_matched([ex_graphs.fn_edge_matched(0)], [{(1, 2): "FN"}], "")

3. Either node matches to a non-split node in the predicted graph

.. jupyter-execute::
    :hide-code:

    plot_matched([ex_graphs.two_to_one(t) for t in [0, 1]], [{4: "NS", (1, 2): "FN"}, {5: "NS", (1, 7): "FN", (2, 3): "FN"}], "")

Intertrack
^^^^^^^^^^

> Intertrack edges connect parent cells to daughter cells.

Reviewer: Added some more detail here. I will note that in the CTC ground truth there are instances of what appears to be the same cell, near the border, connected in consecutive frames with different track IDs (and therefore an intertrack edge), but I don't know if/how we should handle these. Will open an issue.

```suggestion
Intertrack edges connect two nodes with different track IDs, most commonly parent cells to daughter cells, or two cells in non-consecutive frames. Note that intertrack edges are not errors, but support the annotation of Wrong Semantic edges, as detailed below.
```
